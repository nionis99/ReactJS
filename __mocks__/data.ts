export const genres = ['Crime', 'Documentary', 'Horror', 'Comedy'];
export const MovieMockData = {
  id: 0,
  title: 'The Avengers 2',
  vote_average: 7.5,
  release_date: '2012-04-25',
  poster_path: 'https://image.tmdb.org/t/p/w500/cezWGskPY5x7GaglTTRN4Fugfb8.jpg',
  overview: 'Lorem Ipsum is simply dummy text of the printing and typesetting industry.',
  genres: ['Science Fiction', 'Action', 'Adventure'],
  runtime: 143,
};
