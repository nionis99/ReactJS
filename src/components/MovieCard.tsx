import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import moment from 'moment';
import DeleteConfirmation from 'components/Modals/DeleteConfirmation';
import MovieFormModal from 'components/Modals/MovieFormModal';
import { deleteMovie } from 'actions/movieActions';
import { Movie } from 'reducers/movieReducers/types';
import Dots from 'assets/icons/dots.svg';
import XIcon from 'assets/icons/x.svg';
import useStateSelector from '../hooks/useStateSelector';

interface MovieCardProps {
  movie: Movie;
  selectedMovie?: Movie;
  setSelectedMovie: (movie?: Movie) => void;
  onClick: () => void;
}

const MovieCard = ({ movie, selectedMovie, setSelectedMovie, onClick }: MovieCardProps) => {
  const { deleteMovieLoading } = useStateSelector((state) => state.movies);
  const { poster_path, title, release_date, genres } = movie;
  const yearsOfTheMovie = moment(release_date).format('YYYY');
  const [isBlurred, setIsBlurred] = useState(false);
  const [hasMoreActionSelected, setHasMoreActionSelected] = useState(false);
  const [isDeletingMovie, setIsDeletingMovie] = useState(false);
  const [editingMovie, setEditingMovie] = useState<Movie | undefined>();
  const dispatch = useDispatch();

  const onCardMouseLeave = () => {
    if (hasMoreActionSelected) setHasMoreActionSelected(false);
    setIsBlurred(false);
  };

  const onEditMovieClick = (event: React.MouseEvent<HTMLLIElement>) => {
    event.stopPropagation();
    setEditingMovie(movie);
  };

  const onDeleteMovieClick = (event: React.MouseEvent<HTMLLIElement>) => {
    event.stopPropagation();
    setIsDeletingMovie(true);
  };

  const onDeleteMovieConfirmation = () => {
    dispatch(deleteMovie(movie.id, () => setIsDeletingMovie(false)));
    if (selectedMovie) setSelectedMovie(undefined);
  };

  const onMoreActionClose = (event: React.MouseEvent<SVGAElement>) => {
    event.stopPropagation();
    setHasMoreActionSelected(false);
  };

  const onMoreActionClicked = (event: React.MouseEvent<HTMLDivElement>) => {
    event.stopPropagation();
    setHasMoreActionSelected(true);
  };

  return (
    <>
      <div
        className="flex-col cursor-pointer text-white text-opacity-75"
        onMouseEnter={() => setIsBlurred(true)}
        onMouseLeave={onCardMouseLeave}
        onClick={onClick}
      >
        <div className="relative pb-4">
          {isBlurred && (
            <>
              <div className="absolute top-2 right-2" onClick={onMoreActionClicked}>
                <Dots />
              </div>
              {hasMoreActionSelected && (
                <ul className="absolute w-48 right-2 py-1 mt-2 bg-white bg-content rounded shadow-modal">
                  <p className="w-full p-1">
                    <XIcon className="ml-auto mb-1 mr-2 w-3 h-3" onClick={onMoreActionClose} />
                  </p>
                  <li className="px-4 py-2 text-white hover:bg-primary" onClick={onEditMovieClick}>
                    Edit
                  </li>
                  <li className="px-4 py-2 text-white hover:bg-primary" onClick={onDeleteMovieClick}>
                    Delete
                  </li>
                </ul>
              )}
            </>
          )}
          <img src={poster_path} alt={title} />
        </div>
        <p className="flex w-full">
          <span className="font-bold">{title}</span>
          <span className="ml-auto px-2 h-7 py-1 text-xs border-2 rounded-lg border-gray-400 border-opacity-50">
            {yearsOfTheMovie}
          </span>
        </p>
        <small>{genres.join(', ')}</small>
      </div>
      <MovieFormModal
        isOpen={!!editingMovie}
        onClose={() => setEditingMovie(undefined)}
        title="edit movie"
        movie={editingMovie}
      />
      <DeleteConfirmation
        isOpen={isDeletingMovie}
        title="Delete movie"
        description="Are you sure you want to delete this movie?"
        isLoading={deleteMovieLoading}
        onConfirm={onDeleteMovieConfirmation}
        onClose={() => setIsDeletingMovie(false)}
      />
    </>
  );
};

export default MovieCard;
