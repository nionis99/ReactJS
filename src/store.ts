import { createStore, applyMiddleware } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';
import { MovieActionsTypes } from 'actions/movieActions/types';
import rootReducer from 'reducers';
import thunk from 'redux-thunk';

const store = createStore(rootReducer, composeWithDevTools(applyMiddleware(thunk)));

export type RootState = ReturnType<typeof rootReducer>;
export type RootActions = MovieActionsTypes;
export default store;
