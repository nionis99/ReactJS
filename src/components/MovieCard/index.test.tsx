import '@testing-library/jest-dom';
import React from 'react';
import renderer from 'react-test-renderer';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { MovieMockData } from '../../../__mocks__/data';
import MovieCard from '.';
import store from 'store';

describe('Movie card component', () => {
  const onClick = jest.fn();
  const wrapper = (
    <Provider store={store}>
      <BrowserRouter>
        <MovieCard movie={MovieMockData} onClick={onClick} />
      </BrowserRouter>
    </Provider>
  );

  it('should render movie card component', () => {
    const elem = renderer.create(wrapper).toJSON();
    expect(elem).toMatchSnapshot();
  });
});
