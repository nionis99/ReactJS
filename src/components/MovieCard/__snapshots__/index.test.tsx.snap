// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie card component should render movie card component 1`] = `
<div
  className="flex-col cursor-pointer text-white text-opacity-75"
  data-testid={11}
  onClick={[MockFunction]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="relative pb-4"
  >
    <img
      alt="The Avengers 2"
      onError={[Function]}
      src="https://image.tmdb.org/t/p/w500/cezWGskPY5x7GaglTTRN4Fugfb8.jpg"
    />
  </div>
  <p
    className="flex w-full"
  >
    <span
      className="font-bold"
    >
      The Avengers 2
    </span>
    <span
      className="ml-auto px-2 h-7 py-1 text-xs border-2 rounded-lg border-gray-400 border-opacity-50"
    >
      2012
    </span>
  </p>
  <small>
    Science Fiction, Action, Adventure
  </small>
</div>
`;
